2021_scipy_tutorial/fairness-in-AI-systems-instructors.ipynb

```diff
@@ -0,0 +1,4620 @@
+        "\n",
+        "`ThresholdOptimizer` takes in an exisiting (possibly pre-fit) machine learning model whose predictions act as a scoring function and identifies a separate thrceshold for each group in order to optimize some specified objective metric (such as **balanced accuracy**) subject to specified fairness constraints (such as **false negative rate parity**). Thus, the resulting classifier is just a suitably thresholded version of the underlying machinelearning model.\n",
+        "\n",
+        "The constraint **false negative rate parity** requires that all the groups have equal values of false negative rate.\n",
+        "\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "OFOovaN7AwDr"
+      },
+      "source": [
+        "To instatiate our `ThresholdOptimizer`, we pass in:\n",
+        "\n",
+        "*   An existing `estimator` that we wish to threshold. \n",
+        "*   The fairness `constraints` we want to satisfy.\n",
+        "*   The `objective` metric we want to maximize.\n",
+        "\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "8je0grKPWHhy"
+      },
+      "outputs": [],
+      "source": [
+        "# Now we instantite ThresholdOptimizer with the logistic regression estimator\n",
+        "postprocess_est = ThresholdOptimizer(\n",
+        "    estimator=unmitigated_pipeline,\n",
+        "    constraints=\"false_negative_rate_parity\",\n",
+        "    objective=\"balanced_accuracy_score\",\n",
+        "    prefit=True,\n",
+        "    predict_method='predict_proba'\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "VDD86L7eCSe0"
+      },
+      "source": [
+        "In order to use the `ThresholdOptimizer`, we need access to the sensitive features **both during training time and once it's deployed**."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 344
+        },
+        "id": "VCHJBB7x1rAK",
+        "outputId": "50b7e579-3416-4463-faf4-5adf534558b8"
+      },
+      "outputs": [],
+      "source": [
+        "postprocess_est.fit(X_train_bal, Y_train_bal, sensitive_features=A_train_bal)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "YscNZsYU1rCY"
+      },
+      "outputs": [],
+      "source": [
+        "# Record and evaluate the output of the trained ThresholdOptimizer on test data\n",
+        "\n",
+        "Y_pred_postprocess = postprocess_est.predict(X_test, sensitive_features=A_test)\n",
+        "metricframe_postprocess = MetricFrame(\n",
+        "    metrics=metrics_with_err_bars,\n",
```

Comment:
Now that the error bars exercise is optional, should we just have a separate dictionary called `metrics` or `tutorial_metrics` that would just contain all of the metrics that are included in `metrics_to_display`? Then we wouldn't need to use `metrics_to_display` after the error bars section (we would still need to use differences_to_display).

Alternatively, we can keep things as is and I can mention these variables when I go through the MetricFrame section.

Reply:
Honestly, we could just use `metric_dict` in the later sections. Showing `count` isn't super useful in the later sections, so we only need to visualize performance across the three metrics.

Reply:
great. can you implement?

Reply:
[i see below that you just did]